nexus/db-model/src/inventory.rs

```diff
@@ -885,18 +887,205 @@ impl InvPhysicalDisk {
+            .slot_firmware_versions
+            .get(usize::from(firmware.active_slot) - 1)
+        else {
+            return Err(InvNvmeDiskFirmwareError::InvalidActiveSlotFirmware(
+                firmware.active_slot,
+            ));
+        };
+        if let Some(next_active_slot) = firmware.next_active_slot {
+            if !valid_slot.contains(&next_active_slot) {
+                return Err(InvNvmeDiskFirmwareError::InvalidNextActiveSlot(
+                    next_active_slot,
+                ));
+            }
+
+            // next_active_slot maps to a populated version
+            let Some(Some(_fw_string)) = firmware
+                .slot_firmware_versions
+                .get(usize::from(next_active_slot) - 1)
+            else {
+                return Err(
+                    InvNvmeDiskFirmwareError::InvalidNextActiveSlotFirmware(
+                        next_active_slot,
+                    ),
+                );
+            };
+        }
+        // slot fw strings must be a max of 8 bytes and must be ascii characters
+        for fw_string in firmware.slot_firmware_versions.iter().flatten() {
+            if !fw_string.is_ascii() {
+                return Err(InvNvmeDiskFirmwareError::FirmwareVersionNotAscii(
+                    fw_string.clone(),
+                ));
+            }
+            if fw_string.len() > 8 {
+                return Err(InvNvmeDiskFirmwareError::FirmwareVersionTooLong(
+                    fw_string.clone(),
+                ));
+            }
+        }
+
+        Ok(Self {
+            inv_collection_id: inv_collection_id.into(),
+            sled_id: sled_id.into(),
+            slot: sled_slot,
+            active_slot: firmware.active_slot.into(),
+            next_active_slot: firmware.next_active_slot.map(|nas| nas.into()),
+            number_of_slots: firmware.number_of_slots.into(),
+            slot1_is_read_only: firmware.slot1_is_read_only,
+            slot_firmware_versions: firmware.slot_firmware_versions.clone(),
+        })
+    }
+
+    /// Attempt to read the current firmware version.
+    pub fn current_version(&self) -> Option<&str> {
+        match self.active_slot.0 {
+            // be paranoid that we have a value within the NVMe spec
+            slot @ 1..=7 => self
+                .slot_firmware_versions
+                .get(usize::from(slot) - 1)
+                .and_then(|v| v.as_deref()),
+            _ => None,
+        }
+    }
+
+    /// Attempt to read the staged firmware version that will be active upon
+    /// next device reset.
+    pub fn next_version(&self) -> Option<&str> {
+        match self.next_active_slot {
+            // be paranoid that we have a value within the NVMe spec
+            Some(slot) if slot.0 <= 7 && slot.0 >= 1 => self
+                .slot_firmware_versions
+                .get(usize::from(slot.0) - 1)
+                .and_then(|v| v.as_deref()),
+            _ => None,
         }
     }
```

Comment:
These checks should be unnecessary now that the only way to construct a `InvNvmeDiskFirmware` is through the "new" method, but I kind of like the extra safety net.  Thoughts on converting this to `Result<&str, InvNvmeDiskFirmwareError>` and `Result<Option<&str>, InvNvmeDiskFirmwareError>`

Reply:
I think an `Option` is fine -- I'm not sure we'd yet want a caller to treat a `Result::Err` case differently from `None` anyway.